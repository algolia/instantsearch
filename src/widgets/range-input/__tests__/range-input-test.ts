/** @jsx h */

import { render as _originalRender, VNode } from 'preact';
import algoliasearchHelper from 'algoliasearch-helper';
import rangeInput from '../range-input';
import { createSearchClient } from '../../../../test/mock/createSearchClient';

jest.mock('preact', () => {
  const module = jest.requireActual('preact');

  module.render = jest.fn();

  return module;
});
const render = _originalRender as jest.MockedFunction<typeof _originalRender>;

describe('rangeInput', () => {
  const attribute = 'aNumAttr';
  const createContainer = () => document.createElement('div');
  const instantSearchInstance = {};
  const createHelper = () =>
    algoliasearchHelper(createSearchClient(), 'indexName', {
      disjunctiveFacets: [attribute],
    });

  afterEach(() => {
    render.mockReset();
  });

  describe('Usage', () => {
    it('throws without container', () => {
      expect(() => rangeInput({ container: undefined }))
        .toThrowErrorMatchingInlineSnapshot(`
"The \`container\` option is required.

See documentation: https://www.algolia.com/doc/api-reference/widgets/range-input/js/"
`);
    });

    it('is a widget', () => {
      const container = document.createElement('div');
      const widget = rangeInput({ container, attribute: 'price' });

      expect(widget).toEqual(
        expect.objectContaining({
          $$type: 'ais.rangeInput',
          $$widgetType: 'ais.rangeInput',
        })
      );
    });
  });

  describe('Lifecycle', () => {
    describe('dispose', () => {
      it('unmounts the component', () => {
        const container = document.createElement('div');
        const helper = createHelper();
        const widget = rangeInput({
          attribute: 'price',
          container,
        });

        expect(render).toHaveBeenCalledTimes(0);

        widget.dispose({
          state: helper.state,
          helper,
        });

        expect(render).toHaveBeenCalledTimes(1);
        expect(render).toHaveBeenLastCalledWith(null, container);
      });
    });
  });

  it('expect to render with results', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = {
      disjunctiveFacets: [
        {
          name: attribute,
          stats: {
            min: 10,
            max: 500,
          },
        },
      ],
    };

    const widget = rangeInput({
      container,
      attribute,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect((firstRender.props as any).min).toBe(10);
    expect((firstRender.props as any).max).toBe(500);
    expect(firstRender.props).toMatchSnapshot();
  });

  it('expect to render without results', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = [];

    const widget = rangeInput({
      container,
      attribute,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect(firstRender.props).toMatchSnapshot();
  });

  it('expect to render with custom classNames', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = [];

    const widget = rangeInput({
      container,
      attribute,
      cssClasses: {
        root: 'root',
        noRefinement: 'noRefinement',
        form: 'form',
        label: 'label',
        input: 'input',
        inputMin: 'inputMin',
        inputMax: 'inputMax',
        separator: 'separator',
        submit: 'submit',
      },
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect(firstRender.props).toMatchSnapshot();
  });

  it('expect to render with custom templates', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = [];

    const widget = rangeInput({
      container,
      attribute,
      templates: {
        separatorText: 'custom separator',
        submitText: 'custom submit',
      },
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect(firstRender.props).toMatchSnapshot();
  });

  it('expect to render with min', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = [];

    const widget = rangeInput({
      container,
      attribute,
      min: 20,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect((firstRender.props as any).min).toBe(20);
  });

  it('expect to render with max', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = [];

    const widget = rangeInput({
      container,
      attribute,
      max: 480,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect((firstRender.props as any).max).toBe(480);
  });

  it('expect to render with refinement', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = {
      disjunctiveFacets: [
        {
          name: attribute,
          stats: {
            min: 10,
            max: 500,
          },
        },
      ],
    };

    helper.addNumericRefinement(attribute, '>=', 25);
    helper.addNumericRefinement(attribute, '<=', 475);

    const widget = rangeInput({
      container,
      attribute,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect(firstRender.props).toMatchSnapshot();
    expect((firstRender.props as any).values).toEqual({
      min: 25,
      max: 475,
    });
  });

  it('expect to render with refinement at boundaries', () => {
    const container = createContainer();
    const helper = createHelper();
    const results = {};

    helper.addNumericRefinement(attribute, '>=', 10);
    helper.addNumericRefinement(attribute, '<=', 500);

    const widget = rangeInput({
      container,
      attribute,
      min: 10,
      max: 500,
    });

    widget.init({ helper, instantSearchInstance });
    widget.render({ results, helper });

    const firstRender = render.mock.calls[0][0] as VNode;

    expect(render).toHaveBeenCalledTimes(1);
    expect(firstRender.props).toMatchSnapshot();
    expect((firstRender.props as any).values).toEqual({
      min: undefined,
      max: undefined,
    });
  });

  describe('precision', () => {
    it('expect to render with default precision', () => {
      const container = createContainer();
      const helper = createHelper();
      const results = [];

      const widget = rangeInput({
        container,
        attribute,
        precision: 2,
      });

      widget.init({ helper, instantSearchInstance });
      widget.render({ results, helper });

      const firstRender = render.mock.calls[0][0] as VNode;

      expect(render).toHaveBeenCalledTimes(1);
      expect((firstRender.props as any).step).toBe(0.01);
    });

    it('expect to render with precision of 0', () => {
      const container = createContainer();
      const helper = createHelper();
      const results = [];

      const widget = rangeInput({
        container,
        attribute,
        precision: 0,
      });

      widget.init({ helper, instantSearchInstance });
      widget.render({ results, helper });

      const firstRender = render.mock.calls[0][0] as VNode;

      expect(render).toHaveBeenCalledTimes(1);
      expect((firstRender.props as any).step).toBe(1);
    });

    it('expect to render with precision of 1', () => {
      const container = createContainer();
      const helper = createHelper();
      const results = [];

      const widget = rangeInput({
        container,
        attribute,
        precision: 1,
      });

      widget.init({ helper, instantSearchInstance });
      widget.render({ results, helper });

      const firstRender = render.mock.calls[0][0] as VNode;

      expect(render).toHaveBeenCalledTimes(1);
      expect((firstRender.props as any).step).toBe(0.1);
    });
  });
});
