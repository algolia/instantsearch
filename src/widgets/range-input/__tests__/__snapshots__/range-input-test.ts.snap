// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rangeInput expect to render with custom classNames 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form form",
    "input": "ais-RangeInput-input input",
    "inputMax": "ais-RangeInput-input--max inputMax",
    "inputMin": "ais-RangeInput-input--min inputMin",
    "label": "ais-RangeInput-label label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput root",
    "separator": "ais-RangeInput-separator separator",
    "submit": "ais-RangeInput-submit submit",
  },
  "max": 0,
  "min": 0,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "to",
      "submitText": "Go",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;

exports[`rangeInput expect to render with custom templates 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form",
    "input": "ais-RangeInput-input",
    "inputMax": "ais-RangeInput-input--max",
    "inputMin": "ais-RangeInput-input--min",
    "label": "ais-RangeInput-label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput",
    "separator": "ais-RangeInput-separator",
    "submit": "ais-RangeInput-submit",
  },
  "max": 0,
  "min": 0,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "custom separator",
      "submitText": "custom submit",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;

exports[`rangeInput expect to render with refinement 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form",
    "input": "ais-RangeInput-input",
    "inputMax": "ais-RangeInput-input--max",
    "inputMin": "ais-RangeInput-input--min",
    "label": "ais-RangeInput-label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput",
    "separator": "ais-RangeInput-separator",
    "submit": "ais-RangeInput-submit",
  },
  "max": 500,
  "min": 10,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "to",
      "submitText": "Go",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": 475,
    "min": 25,
  },
}
`;

exports[`rangeInput expect to render with refinement at boundaries 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form",
    "input": "ais-RangeInput-input",
    "inputMax": "ais-RangeInput-input--max",
    "inputMin": "ais-RangeInput-input--min",
    "label": "ais-RangeInput-label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput",
    "separator": "ais-RangeInput-separator",
    "submit": "ais-RangeInput-submit",
  },
  "max": 500,
  "min": 10,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "to",
      "submitText": "Go",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;

exports[`rangeInput expect to render with results 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form",
    "input": "ais-RangeInput-input",
    "inputMax": "ais-RangeInput-input--max",
    "inputMin": "ais-RangeInput-input--min",
    "label": "ais-RangeInput-label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput",
    "separator": "ais-RangeInput-separator",
    "submit": "ais-RangeInput-submit",
  },
  "max": 500,
  "min": 10,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "to",
      "submitText": "Go",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;

exports[`rangeInput expect to render without results 1`] = `
Object {
  "cssClasses": Object {
    "form": "ais-RangeInput-form",
    "input": "ais-RangeInput-input",
    "inputMax": "ais-RangeInput-input--max",
    "inputMin": "ais-RangeInput-input--min",
    "label": "ais-RangeInput-label",
    "noRefinement": "ais-RangeInput--noRefinement",
    "root": "ais-RangeInput",
    "separator": "ais-RangeInput-separator",
    "submit": "ais-RangeInput-submit",
  },
  "max": 0,
  "min": 0,
  "refine": [Function],
  "step": 1,
  "templateProps": Object {
    "templates": Object {
      "separatorText": "to",
      "submitText": "Go",
    },
    "templatesConfig": undefined,
    "useCustomCompileOptions": Object {
      "separatorText": true,
      "submitText": true,
    },
  },
  "values": Object {
    "max": undefined,
    "min": undefined,
  },
}
`;
